File: Assets/felucca/src/felucca/components/following_camera.clj
(ns felucca.components.following-camera
  (:require [arcadia.core :refer [defmutable log update-state state]]
            [felucca.vector3 :as v3])
  (:import [UnityEngine Vector3]))

(defmutable FollowingCamera [^GameObject followee
                             ^Vector3 followee-offset])

(defn start! [^GameObject game-object role-key]
  (update-state game-object
                :following-camera
                (fn [{:keys [followee] :as following-camera-state}]
                  (assoc following-camera-state
                         :followee-offset
                         (v3/v- (.. game-object transform position)
                                (.. followee transform position))))))

(defn late-update! [^GameObject game-object role-key]
  (let [following-camera-state (state game-object :following-camera)]
    (set! (.. game-object transform position)
          (.. (:followee following-camera-state) transform position))))

(def hooks
  {:start #'start!
   :late-update #'late-update!})

File: Assets/Scripts/Felucca/Components/FollowingCamera.cs
using System;
using UnityEngine;

namespace Felucca.Components {
    public class FollowingCamera : MonoBehaviour {
        public Camera mainCamera;

        public Creature followee;
        public Vector3  position;
        public Vector3  rotation;
        public bool     isFollowing;

        private void Start() {
            position = new Vector3(0, 43, 0);
            rotation = new Vector3(45, -45, 0);

            mainCamera = gameObject.GetComponent<Camera>();
            mainCamera.orthographic = true;
            mainCamera.orthographicSize = 10f;
            mainCamera.backgroundColor = Color.clear;
            mainCamera.clearFlags = CameraClearFlags.SolidColor;

            mainCamera.transform.localPosition = position;
            mainCamera.transform.localEulerAngles = rotation;
        }

        private void LateUpdate() {
            if (followee != null && isFollowing) {
                // FIXME: where are these magic numbers coming from? They are
                // likely related to the camera position y axis.
                position.x = followee.transform.position.x + 30;
                position.z = followee.transform.position.z - 30;
            }

            mainCamera.transform.localPosition = position;
            mainCamera.transform.localEulerAngles = rotation;
        }

        public void Follow(Creature creature) {
            followee = creature;
        }
    }
}
